src/client_backend/dynamic_grpc/dynamic_grpc_client.h

```diff
@@ -0,0 +1,178 @@
+  // If this parameter is empty, the default roots will be used. The
+  // default roots can be overridden using the
+  // GRPC_DEFAULT_SSL_ROOTS_FILE_PATH environment variable pointing
+  // to a file on the file system containing the roots.
+  std::string root_certificates;
+  // File containing the PEM encoding of the client's private key.
+  // This parameter can be empty if the client does not have a
+  // private key.
+  std::string private_key;
+  // File containing the PEM encoding of the client's certificate chain.
+  // This parameter can be empty if the client does not have a
+  // certificate chain.
+  std::string certificate_chain;
+};
+
+class DynamicGrpcClient;
+
+//==============================================================================
+// An DynamicGrpcRequest represents an inflght request on gRPC.
+//
+class DynamicGrpcRequest {
+ public:
+  DynamicGrpcRequest(OnCompleteFn callback = nullptr)
+      : callback_(callback), grpc_status_()
+  {
+  }
+
+  tc::RequestTimers& Timer() { return timer_; }
+  friend DynamicGrpcClient;
+
+ private:
+  OnCompleteFn callback_;
+  // Variables for GRPC call
+  grpc::ClientContext grpc_context_;
+  grpc::Status grpc_status_;
+
+  // The timers for infer request.
+  tc::RequestTimers timer_;
+};
+
+//==============================================================
+///
+class DynamicGrpcInferResult : public InferResult {
+ public:
+  DynamicGrpcInferResult(bool request_status) : request_status_(request_status)
+  {
+  }
+
+  /// See InferResult::Id()
+  Error Id(std::string* id) const override;
+  /// See InferResult::RequestStatus()
+  Error RequestStatus() const override;
+  /// See InferResult::RawData()
+  Error RawData(
+      const std::string& output_name, std::vector<uint8_t>& buf) const override;
+
+ private:
+  bool request_status_;
+};
+
+//==============================================================================
+class DynamicGrpcClient {
+ public:
+  DynamicGrpcClient(
+      const std::string& url, const std::string& grpc_method, bool verbose,
+      bool use_ssl, const SslOptions& ssl_options);
+
+  ~DynamicGrpcClient();
+
+  /// Runs an synchronous inference over gRPC bi-directional streaming API.
+  /// A stream must be established with a call to StartStream() before calling
+  /// this function.
+  /// \param result Returns the result of inference.
+  /// \param options The options for inference request.
+  /// \param inputs The vector of InferInput describing the model inputs.
+  /// \param outputs Optional vector of InferRequestedOutput describing how the
+  /// output must be returned. If not provided then all the outputs in the model
+  /// config will be returned as default settings.
+  /// \param headers Optional map specifying additional HTTP headers to include
+  /// in the metadata of gRPC request.
+  /// \return Error object indicating success or failure of the
+  /// request.
+  Error BidiStreamRPC(
```

Comment:
I think including "inference" in the function name might be more understandable for a new reader.

The fact that it's using a bidirectional stream RPC is an implementation detail, but this class is intended to send inference requests (messages), essentially.

Reply:
I would consider "inference" as a concept that exists higher above the dynamic grpc client, and dynamic grpc client doesn't really know/care how it's being used or called. The goal of dynamic grpc client is to support all of the four grpc method types in a generic manner:
- Unary RPC
- Client Streaming RPC
- Server Streaming RPC
- Bidirectional Streaming RPC

For the current phase, we are only introducing the bidirectional one since that's what the stakeholders need, but I foresee the dynamic grpc client to support the rest through separate APIs (in theory, BidiStreamRPC could handle other ones but some stakeholders may not want to establish stream for simple RPC calls).

Reply:
Hmm, I guess that's fair. But then shouldn't ever mention of inference concept be scrubbed from the class? There's still mentions of inference in documentation and UpdateInferStat.

I'll resolve this thread, but something to think about for future, what levels of abstraction we want to have exposed to this level.